src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonNode.cs

```diff
@@ -232,6 +234,139 @@ public JsonNode Root
+            {
+                return jObject.DeepCloneObject();
+            }
+            else if (this is JsonArray jArray)
+            {
+                return jArray.DeepCloneArray();
+            }
+            else
+            {
+                return AsValue().DeepCloneValue();
+            }
+        }
+
+        /// <summary>
+        /// Returns <see cref="JsonValueKind"/> of current instance.
+        /// </summary>
+        public JsonValueKind GetValueKind()
+        {
+            if (this is JsonObject)
+            {
+                return JsonValueKind.Object;
+            }
+            else if (this is JsonArray)
+            {
+                return JsonValueKind.Array;
+            }
+            else
+            {
+                return AsValue().GetInternalValueKind();
+            }
+        }
+
+        /// <summary>
+        /// Returns property name from parent object.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// The current parent is not a <see cref="JsonObject"/>
+        /// </exception>
+        public string GetPropertyName()
+        {
+            if (_parent is JsonObject jObject)
+            {
+                return jObject.GetPropertyName(this);
+            }
+
+            throw new InvalidOperationException(SR.Format(SR.NodeWrongType, nameof(JsonObject)));
+        }
+
+        /// <summary>
+        /// Returns index from parent array.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// The current parent is not a <see cref="JsonArray"/>
+        /// </exception>
+        public int GetElementIndex()
+        {
+            if (_parent is JsonArray jArray)
+            {
+                return jArray.GetElementIndex(this);
+            }
+
+            throw new InvalidOperationException(SR.Format(SR.NodeWrongType, nameof(JsonArray)));
+        }
+
+        /// <summary>
+        /// Compares the values of two nodes, including the values of all descendant nodes.
+        /// </summary>
+        /// <param name="node1">The <see cref="JsonNode"/> to compare.</param>
+        /// <param name="node2">The <see cref="JsonNode"/> to compare.</param>
+        /// <returns><c>true</c> if the tokens are equal; otherwise <c>false</c>.</returns>
+        public static bool DeepEquals(JsonNode? node1, JsonNode? node2)
+        {
+            if (node1 is null && node2 is null)
+            {
+                return true;
+            }
+
+            if (node1 is null && node2 is not null)
+            {
+                return false;
+            }
+
+            if (node1 is JsonObject jsonObject)
+            {
+                return jsonObject.DeepEquals(node2);
+            }
+
+            if (node1 is JsonArray jsonArray)
+            {
+                return jsonArray.DeepEquals(node2);
+            }
+
+            Debug.Assert(node1 is not null);
+            return node1.AsValue().DeepEquals(node2);
```

Comment:
It seems to me that all these if checks are unnecessary since it's always calling the same virtual method:
```suggestion
            return node1.DeepEquals(node2);
```